types/src/contract_messages/messages.rs

```diff
@@ -0,0 +1,324 @@
+)]
+pub struct MessageChecksum(
+    #[cfg_attr(feature = "json-schema", schemars(skip, with = "String"))]
+    pub  [u8; MESSAGE_CHECKSUM_LENGTH],
+);
+
+impl MessageChecksum {
+    /// Returns inner value of the message checksum.
+    pub fn value(&self) -> [u8; MESSAGE_CHECKSUM_LENGTH] {
+        self.0
+    }
+
+    /// Formats the `MessageChecksum` as a human readable string.
+    pub fn to_formatted_string(self) -> String {
+        format!(
+            "{}{}",
+            MESSAGE_CHECKSUM_STRING_PREFIX,
+            base16::encode_lower(&self.0),
+        )
+    }
+
+    /// Parses a string formatted as per `Self::to_formatted_string()` into a
+    /// `MessageChecksum`.
+    pub fn from_formatted_str(input: &str) -> Result<Self, FromStrError> {
+        let hex_addr = input
+            .strip_prefix(MESSAGE_CHECKSUM_STRING_PREFIX)
+            .ok_or(FromStrError::InvalidPrefix)?;
+
+        let bytes =
+            <[u8; MESSAGE_CHECKSUM_LENGTH]>::try_from(checksummed_hex::decode(hex_addr)?.as_ref())?;
+        Ok(MessageChecksum(bytes))
+    }
+}
+
+impl ToBytes for MessageChecksum {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut buffer = bytesrepr::allocate_buffer(self)?;
+        buffer.append(&mut self.0.to_bytes()?);
+        Ok(buffer)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.0.serialized_length()
+    }
+}
+
+impl FromBytes for MessageChecksum {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (checksum, rem) = FromBytes::from_bytes(bytes)?;
+        Ok((MessageChecksum(checksum), rem))
+    }
+}
+
+impl Serialize for MessageChecksum {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        if serializer.is_human_readable() {
+            self.to_formatted_string().serialize(serializer)
+        } else {
+            self.0.serialize(serializer)
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for MessageChecksum {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        if deserializer.is_human_readable() {
+            let formatted_string = String::deserialize(deserializer)?;
+            MessageChecksum::from_formatted_str(&formatted_string).map_err(SerdeError::custom)
+        } else {
+            let bytes = <[u8; MESSAGE_CHECKSUM_LENGTH]>::deserialize(deserializer)?;
+            Ok(MessageChecksum(bytes))
+        }
+    }
+}
+
+const MESSAGE_PAYLOAD_TAG_LENGTH: usize = U8_SERIALIZED_LENGTH;
+
+/// Tag for a message payload that contains a human readable string.
+pub const MESSAGE_PAYLOAD_STRING_TAG: u8 = 0;
+
+/// The payload of the message emitted by an addressable entity during execution.
+#[derive(Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
+#[cfg_attr(feature = "datasize", derive(DataSize))]
+#[cfg_attr(feature = "json-schema", derive(JsonSchema))]
+pub enum MessagePayload {
```

Comment:
What other variants could be added here in the future, and why CLValue wasn't considered for this purpose? If there are no other variants planned, why not just make it a new type over String?

Reply:
As discussed offline future variants could have richer structure so it's not really a subset or replacement of a CLValue